core/sr-primitives/src/sr_arithmetic.rs

```diff
@@ -669,16 +669,824 @@ impl CheckedAdd for Fixed64 {
+						r.set(d, v);
+					}
+					r.set(n - 1, self_norm.get(n - 1) >> normalizer_bits);
+				} else {
+					r = self_norm;
+				}
+			}
+
+			Some((q, r))
+		}
+	}
+
+	#[cfg(feature = "std")]
+	impl rstd::fmt::Debug for Number{
+		fn fmt(&self, f: &mut rstd::fmt::Formatter<'_>) -> rstd::fmt::Result {
+			write!(
+				f,
+				"Number {{ {:?} ({:?})}}",
+				self.digits,
+				u128::try_from(self.clone()).unwrap_or_else(|_| 0),
+			)
+		}
+	}
+
+	impl PartialEq for Number {
+		fn eq(&self, other: &Self) -> bool {
+			// sadly, we have to reallocate here as strip mutably uses self.
+			let mut lhs = self.clone();
+			let mut rhs = other.clone();
+			lhs.lstrip();
+			rhs.lstrip();
+			lhs.digits.eq(&rhs.digits)
+		}
+	}
+
+	impl Eq for Number {}
+
+	impl Ord for Number {
+		fn cmp(&self, other: &Self) -> Ordering {
+			// sadly, we have to reallocate here as strip mutably uses self.
+			let mut lhs = self.clone();
+			let mut rhs = other.clone();
+			// strip both
+			lhs.lstrip();
+			rhs.lstrip();
+			// then resize to max
+			lhs.resize(&mut rhs);
+			// then we can iterate safely
+			let mut r = Ordering::Equal;
+			for (i, p) in lhs.digits.iter().enumerate() {
+				let q = rhs.digits[i];
+				match p.cmp(&q) {
+					Ordering::Less => { r = Ordering::Less; break} ,
+					Ordering::Greater => { r = Ordering::Greater; break} ,
+					_ => continue,
+				}
+			}
+			r
+		}
+	}
+
+	impl PartialOrd for Number {
+		fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+			Some(self.cmp(other))
+		}
+	}
+
+	impl ops::Add for Number {
+		type Output = Self;
+		fn add(self, rhs: Self) -> Self::Output {
+			self.add(rhs)
+		}
+	}
+
+	impl ops::Sub for Number {
+		type Output = Self;
+		fn sub(self, rhs: Self) -> Self::Output {
+			self.sub(rhs).unwrap_or_else(|e| e)
+		}
+	}
+
+	impl ops::Mul for Number {
+		type Output = Self;
+		fn mul(self, rhs: Self) -> Self::Output {
+			self.mul(rhs)
+		}
+	}
+
+	impl Zero for Number {
+		fn zero() -> Self {
+			Self { digits: vec![Zero::zero()] }
+		}
+
+		fn is_zero(&self) -> bool {
+			self.digits.iter().all(|d| d.is_zero())
+		}
+	}
+
+	macro_rules! impl_try_from_number_for {
+		($([$type:ty, $len:expr]),+) => {
+			$(
+				impl TryFrom<Number> for $type {
+					type Error = &'static str;
+					fn try_from(mut value: Number) -> Result<$type, Self::Error> {
+						value.lstrip();
+						let error_message = concat!("cannot fit a number into ", stringify!($type));
+						if value.len() * SHIFT > $len {
+							Err(error_message)
+						} else {
+							let checked_pow = |a: $type, b: u32| -> Option<$type> {
+								let mut acc: $type = One::one();
+								for _ in 0..b {
+									acc = acc.checked_mul(a)?;
+								}
+								Some(acc)
+							};
+							let mut acc: $type = Zero::zero();
+							for (i, d) in value.digits.iter().rev().cloned().enumerate() {
+								let d: $type = d.into();
+								let b: $type = B.into();
+								let m = checked_pow(b, i as u32).ok_or(error_message)?;
```

Comment:
Can't you just use left shifts here since B is a power of two? Also, I don't think the overflow checking is necessary since you already know `value.len() * SHIFT <= $len` with `$len` as a multiple of `SHIFT`.

Reply:
ahh right this is a dichotomy actually in the code: I started with `d` value being any to it being the more optimised `leading_zeros()` (defined at the top of the function) wich can then be used with shifts. POwer can go away now 👍

Reply:
Okay sorry my previous comment was confused and was about something else. 

Does B matter here? I can optimize the closure to shift if `i` is a power of two though

Reply:
I don't think you even need the variable `i`. Just like

```rust
let mut acc: $type = Zero::zero();
for d in value.digits.iter().rev().cloned() {
    let d: $type = d.into();
    acc = (acc << SHIFT) + d;
}
```

Reply:
implemented with a bit of change